src/Components/Components/src/Routing/Router.cs

```diff
@@ -79,41 +86,57 @@ protected virtual void Render(RenderTreeBuilder builder, Type handler, IDictiona
         {
             var locationPath = UriHelper.ToBaseRelativePath(_baseUri, _locationAbsolute);
             locationPath = StringUntilAny(locationPath, _queryOrHashStartChar);
             var context = new RouteContext(locationPath);
             Routes.Route(context);
 
-            if (context.Handler == null)
+            if (context.Handler != null)
+            {
+                if (!typeof(IComponent).IsAssignableFrom(context.Handler))
+                {
+                    throw new InvalidOperationException($"The type {context.Handler.FullName} " +
+                        $"does not implement {typeof(IComponent).FullName}.");
+                }
+
+                _renderHandle.Render(builder => Render(builder, context.Handler, context.Parameters));
+            }
+            else
             {
-                if (FallbackComponent != null)
+                if (!isNavigationIntercepted && NotFoundContent != null)
                 {
-                    context.Handler = FallbackComponent;
+                    // We did not find a Component that matches the route.
+                    // Only show the NotFoundContent if the application developer programatically got us here i.e we did not
+                    // intercept the navigation. In all other cases, force a browser navigation since this could be non-Blazor content.
+                    _renderHandle.Render(NotFoundContent);
                 }
                 else
                 {
-                    throw new InvalidOperationException($"'{nameof(Router)}' cannot find any component with a route for '/{locationPath}', and no fallback is defined.");
+                    UriHelper.NavigateTo(_locationAbsolute, forceLoad: true);
                 }
             }
+        }
 
-            if (!typeof(IComponent).IsAssignableFrom(context.Handler))
+        private void OnLocationChanged(object sender, LocationChangedEventArgs args)
+        {
+            _locationAbsolute = args.Location;
+            if (_renderHandle.IsInitialized && Routes != null)
             {
-                throw new InvalidOperationException($"The type {context.Handler.FullName} " +
-                    $"does not implement {typeof(IComponent).FullName}.");
+                Refresh(args.IsNavigationIntercepted);
             }
-
-            _renderHandle.Render(builder => Render(builder, context.Handler, context.Parameters));
         }
 
-        private void OnLocationChanged(object sender, string newAbsoluteUri)
+        Task IHandleAfterRender.OnAfterRenderAsync()
         {
-            _locationAbsolute = newAbsoluteUri;
-            if (_renderHandle.IsInitialized)
+            if (!_navigationInterceptionEnabled && ComponentContext.IsConnected)
             {
-                Refresh();
+                _navigationInterceptionEnabled = true;
+                return JSRuntime.InvokeAsync<object>(Interop.EnableNavigationInterception);
```

Comment:
This is a bit awkward - sorry I didn't anticipate this when we discussed the design.

In general the intent is for the core `Microsoft.AspNetCore.Components` package not to make assumptions about the UI rendering technology. As such it doesn't assume that the UI renderer is a browser that would even have an associated JS runtime. The concept of "web browser" of that is in higher layers (e.g., `Microsoft.AspNetCore.Components.Browser`). The reason for this is to keep open the possibility that we allow completely different UI rendering technologies, such as native apps. So, AFAIK (and please correct me if I'm wrong), there aren't any other places in `Microsoft.AspNetCore.Components` where the framework makes JS interop calls.

To avoid making a JS interop call directly from here, we might need to add a new interface method. We could either:

 * Put a new method, `EnableNavigationInterception`, onto `IUriHelper` (since there are different implementations of that in different environments). It might be a bit weird for this to be public, though.
 * Or make a new internal DI service for enabling navigation interception. It could be a new interface `INavigationInterception` where the existing concrete `IUriHelper` implementations also implement that, so there's not much extra code or config to implement. The new interface would just be to segregate it from `IUriHelper` so it doesn't pollute intellisense.

Sorry that's extra hassle. Do you think this is viable?